superset/assets/src/explore/components/controls/MetricsControl.jsx

```diff
@@ -60,6 +60,25 @@ function isDictionaryForAdhocMetric(value) {
   return value && !(value instanceof AdhocMetric) && value.expressionType;
 }
 
+function columnsContainMetric(metric, columns) {
+  if (metric) {
+    const name = metric.column ? metric.column.column_name : metric.column_name || metric;
+    const columnNames = columns.map(column => (column.column_name || column.metric_name));
+    if (name && columnNames.indexOf(name) < 0) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function columnsContainAllMetrics(value, nextProps) {
+  const allColumns = [...nextProps.columns, ...nextProps.savedMetrics];
```

Comment:
Can create a lookup for all columns first and probably better chaining.

```js
const columnNames = new Set(
  [...nextProps.columns, ...nextProps.savedMetrics]
    .map(({ column_name, metric_name }) => (column_name || metric_name))
);

return (Array.isArray(value) ? value : [value])
  // filter null/undefined/empty metric
  .filter(metric => metric)
  // convert to metric name
  .map(metric => metric.column ? metric.column.column_name : metric.column_name || metric)
  // filter null/undefined/empty name
  .filter(name => name)
  // check if the valid names are in the columns
  .every(name => columnNames.has(name));
```

Reply:
thanks fixed!